### PR TITLE
Add compconfig files

### PR DESCRIPTION
**Pull request by jdebacker**
This PR adds compconfig files to help run CCC on the COMP platform.

### CODE REVIEW

**Comment by jdebacker** [comment]
@hdoupe Let me know how this looks.  I have not included any `__init__.py` files - do I need to?

**Comment by hdoupe** [comment]
Yeah, the `__init__.py` files are necessary. The directory structure should look something like this:
```bash
compconfig/
├── compconfig
│   ├── functions.py
│   ├── __init__.py
│   └── tests
│       ├── __init__.py
│       └── test_functions.py
├── install.sh
└── setup.py
```

The `cdk-init` bash command should write these files for you.

**Comment by jdebacker** [comment]
Thanks @hdoupe.  I added `__init__.py` files and reorganized the `compconfig` directory.  Let me know if this is right.

**Comment by hdoupe** [comment]
Thanks, there are a couple more changes::
- There should be a `setup.py` file that looks just like [this][1].
- Drop extra function call [here][2].
- `compconfig/compconfig/__init__.py` should have this at the top: `from .functions import *`.

The `cdk-init` command should have written these files. Did the command work for you? If so, which version of `compdevkit` do you have?

[1]: https://github.com/hdoupe/Matchups/blob/master/compconfig/setup.py
[2]: https://github.com/jdebacker/Cost-of-Capital-Calculator/blob/comp2/compconfig/compconfig/tests/test_functions.py#L21

**Comment by jdebacker** [comment]
@hdoupe 

```
compdevkit                1.3.1                    pypi_0    pypi
```

This did not write the `__init__.py` files.  It did write a `functions.py`, but it was empty (except for one line comment saying to insert code` and it did write a `setup.py` script - I just neglected to commit that.

Fixing now...

**Comment by hdoupe** [comment]
> This did not write the __init__.py files. It did write a functions.py, but it was empty (except for one line comment saying to insert codeand it did write asetup.py` script - I just neglected to commit that.

Ah, I see. Thanks for the info.

> Fixing now...

Sweet, thanks.

**Comment by hdoupe** [comment]
OK, we are almost there. In `tests/test_functions.py`, change this:

```python
from compdevkit import FunctionsTest
import sys
import os.path
sys.path.append(
    os.path.abspath(os.path.join(os.path.dirname(__file__),
                                 os.path.pardir)))
import functions

```

to this:

```python
from compdevkit import FunctionsTest

from compconfig import functions

```

**Comment by jdebacker** [comment]
In local testing, getting this error:

```
(ccc-dev) A02655JDEBACKER:compconfig jason.debacker$ pytest
============================= test session starts ==============================
platform darwin -- Python 3.6.8, pytest-4.4.1, py-1.8.0, pluggy-0.9.0
rootdir: /Users/jason.debacker/repos/CCC, inifile: setup.cfg
plugins: xdist-1.26.1, pep8-1.0.6, forked-1.0.2
collected 1 item

compconfig/tests/test_functions.py F                                     [100%]

=================================== FAILURES ===================================
_____________________________ test_get_parameters ______________________________

self = <compdevkit.FunctionsTest object at 0x1227f26a0>

    def test_get_inputs(self):
        init_metaparams, init_modparams = self.get_inputs({})

        try:
>           json.dumps(init_metaparams)

/anaconda3/envs/ccc-dev/lib/python3.6/site-packages/compdevkit/__init__.py:95:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

obj = OrderedDict([('year', {'title': 'Start year', 'value': [OrderedDict([('value', 2019)])], 'type': 'int', 'number_dims':...tors': OrderedDict([('range', OrderedDict([('max', 2029), ('min', 2015)]))]), 'description': 'Year for parameters.'})])
skipkeys = False, ensure_ascii = True, check_circular = True, allow_nan = True
cls = None, indent = None, separators = None, default = None

    def dumps(obj, *, skipkeys=False, ensure_ascii=True, check_circular=True,
            allow_nan=True, cls=None, indent=None, separators=None,
            default=None, sort_keys=False, **kw):
        """Serialize ``obj`` to a JSON formatted ``str``.

        If ``skipkeys`` is true then ``dict`` keys that are not basic types
        (``str``, ``int``, ``float``, ``bool``, ``None``) will be skipped
        instead of raising a ``TypeError``.

        If ``ensure_ascii`` is false, then the return value can contain non-ASCII
        characters if they appear in strings contained in ``obj``. Otherwise, all
        such characters are escaped in JSON strings.

        If ``check_circular`` is false, then the circular reference check
        for container types will be skipped and a circular reference will
        result in an ``OverflowError`` (or worse).

        If ``allow_nan`` is false, then it will be a ``ValueError`` to
        serialize out of range ``float`` values (``nan``, ``inf``, ``-inf``) in
        strict compliance of the JSON specification, instead of using the
        JavaScript equivalents (``NaN``, ``Infinity``, ``-Infinity``).

        If ``indent`` is a non-negative integer, then JSON array elements and
        object members will be pretty-printed with that indent level. An indent
        level of 0 will only insert newlines. ``None`` is the most compact
        representation.

        If specified, ``separators`` should be an ``(item_separator, key_separator)``
        tuple.  The default is ``(', ', ': ')`` if *indent* is ``None`` and
        ``(',', ': ')`` otherwise.  To get the most compact JSON representation,
        you should specify ``(',', ':')`` to eliminate whitespace.

        ``default(obj)`` is a function that should return a serializable version
        of obj or raise TypeError. The default simply raises TypeError.

        If *sort_keys* is true (default: ``False``), then the output of
        dictionaries will be sorted by key.

        To use a custom ``JSONEncoder`` subclass (e.g. one that overrides the
        ``.default()`` method to serialize additional types), specify it with
        the ``cls`` kwarg; otherwise ``JSONEncoder`` is used.

        """
        # cached encoder
        if (not skipkeys and ensure_ascii and
            check_circular and allow_nan and
            cls is None and indent is None and separators is None and
            default is None and not sort_keys and not kw):
>           return _default_encoder.encode(obj)

/anaconda3/envs/ccc-dev/lib/python3.6/json/__init__.py:231:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <json.encoder.JSONEncoder object at 0x10e993080>
o = OrderedDict([('year', {'title': 'Start year', 'value': [OrderedDict([('value', 2019)])], 'type': 'int', 'number_dims':...tors': OrderedDict([('range', OrderedDict([('max', 2029), ('min', 2015)]))]), 'description': 'Year for parameters.'})])

    def encode(self, o):
        """Return a JSON string representation of a Python data structure.

        >>> from json.encoder import JSONEncoder
        >>> JSONEncoder().encode({"foo": ["bar", "baz"]})
        '{"foo": ["bar", "baz"]}'

        """
        # This is for extremely simple cases and benchmarks.
        if isinstance(o, str):
            if self.ensure_ascii:
                return encode_basestring_ascii(o)
            else:
                return encode_basestring(o)
        # This doesn't pass the iterator directly to ''.join() because the
        # exceptions aren't as detailed.  The list call should be roughly
        # equivalent to the PySequence_Fast that ''.join() would do.
>       chunks = self.iterencode(o, _one_shot=True)

/anaconda3/envs/ccc-dev/lib/python3.6/json/encoder.py:199:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <json.encoder.JSONEncoder object at 0x10e993080>
o = OrderedDict([('year', {'title': 'Start year', 'value': [OrderedDict([('value', 2019)])], 'type': 'int', 'number_dims':...tors': OrderedDict([('range', OrderedDict([('max', 2029), ('min', 2015)]))]), 'description': 'Year for parameters.'})])
_one_shot = True

    def iterencode(self, o, _one_shot=False):
        """Encode the given object and yield each string
        representation as available.

        For example::

            for chunk in JSONEncoder().iterencode(bigobject):
                mysocket.write(chunk)

        """
        if self.check_circular:
            markers = {}
        else:
            markers = None
        if self.ensure_ascii:
            _encoder = encode_basestring_ascii
        else:
            _encoder = encode_basestring

        def floatstr(o, allow_nan=self.allow_nan,
                _repr=float.__repr__, _inf=INFINITY, _neginf=-INFINITY):
            # Check for specials.  Note that this type of test is processor
            # and/or platform-specific, so do tests which don't depend on the
            # internals.

            if o != o:
                text = 'NaN'
            elif o == _inf:
                text = 'Infinity'
            elif o == _neginf:
                text = '-Infinity'
            else:
                return _repr(o)

            if not allow_nan:
                raise ValueError(
                    "Out of range float values are not JSON compliant: " +
                    repr(o))

            return text


        if (_one_shot and c_make_encoder is not None
                and self.indent is None):
            _iterencode = c_make_encoder(
                markers, self.default, _encoder, self.indent,
                self.key_separator, self.item_separator, self.sort_keys,
                self.skipkeys, self.allow_nan)
        else:
            _iterencode = _make_iterencode(
                markers, self.default, _encoder, self.indent, floatstr,
                self.key_separator, self.item_separator, self.sort_keys,
                self.skipkeys, _one_shot)
>       return _iterencode(o, 0)

/anaconda3/envs/ccc-dev/lib/python3.6/json/encoder.py:257:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <json.encoder.JSONEncoder object at 0x10e993080>, o = 2019

    def default(self, o):
        """Implement this method in a subclass such that it returns
        a serializable object for ``o``, or calls the base implementation
        (to raise a ``TypeError``).

        For example, to support arbitrary iterators, you could
        implement default like this::

            def default(self, o):
                try:
                    iterable = iter(o)
                except TypeError:
                    pass
                else:
                    return list(iterable)
                # Let the base class default method raise the TypeError
                return JSONEncoder.default(self, o)

        """
        raise TypeError("Object of type '%s' is not JSON serializable" %
>                       o.__class__.__name__)
E       TypeError: Object of type 'int64' is not JSON serializable

/anaconda3/envs/ccc-dev/lib/python3.6/json/encoder.py:180: TypeError

During handling of the above exception, another exception occurred:

    def test_get_parameters():
        ta = FunctionsTest(
            get_inputs=functions.get_inputs,
            validate_inputs=functions.validate_inputs,
            run_model=functions.run_model,
            ok_adjustment={"ccc": {"CIT_rate": 0.21}},
            bad_adjustment={"ccc": {"CIT_rate": -0.1}}
        )
>       ta.test()

compconfig/tests/test_functions.py:14:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/anaconda3/envs/ccc-dev/lib/python3.6/site-packages/compdevkit/__init__.py:87: in test
    self.test_get_inputs()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <compdevkit.FunctionsTest object at 0x1227f26a0>

    def test_get_inputs(self):
        init_metaparams, init_modparams = self.get_inputs({})

        try:
            json.dumps(init_metaparams)
        except TypeError as e:
            raise SerializationError(
>               (f"Meta parameters must be JSON serializable: \n\n\t{str(e)}\n"
                 f"\nHint: try setting `serializable=True` in `Parameters.specification`.")
            )
E           compdevkit.exceptions.SerializationError: Meta parameters must be JSON serializable:
E
E           	Object of type 'int64' is not JSON serializable
E
E           Hint: try setting `serializable=True` in `Parameters.specification`.

/anaconda3/envs/ccc-dev/lib/python3.6/site-packages/compdevkit/__init__.py:98: SerializationError
=========================== 1 failed in 2.35 seconds ===========================
```

There is no kwarg `serialization` to set to `True` in `parameters.Specifications()`.

**Comment by hdoupe** [comment]
Starting in [ParamTools 0.5.3](https://github.com/PSLmodels/ParamTools/releases/tag/0.5.3) there's a `serializable` flag.

File: compconfig/compconfig/functions.py [hunk 1]
```diff
@@ -30,9 +34,10 @@ def get_inputs(meta_params_dict):
     meta_params = MetaParams()
     meta_params.adjust(meta_params_dict)
     params = Specifications()
+    params.specification(serializable=True)
```
Comment (jdebacker) [inline compconfig/compconfig/functions.py]: This and another calls to params.specifications()` in other functions doesn't seem to be working to help me pass the tests.  Still getting error that json not serializable.

File: compconfig/compconfig/functions.py [hunk 2]
```diff
@@ -30,9 +34,10 @@ def get_inputs(meta_params_dict):
...
     spec = params.specification(
         meta_data=True,
-        year=meta_params.start_year
+        year=meta_params.year
```
Comment (hdoupe) [inline compconfig/compconfig/functions.py]:
```suggestion
        serializable=True,
        year=meta_params.year
```

File: compconfig/compconfig/functions.py (hunk 1 above)
Comment (hdoupe) [reply compconfig/compconfig/functions.py]: The specification method doesn't change the `params` object. You have to use it like this:
```python
spec = params.specification(serializable=True, meta_data=True, **otherargs)
```

File: compconfig/compconfig/functions.py (hunk 2 above)
Comment (jdebacker) [reply compconfig/compconfig/functions.py]: Ok- tried this and still failing.  Do I need something similar in other functions?  Others create an instance of `Specifications`, but do not then create an object  named `spec`...

File: compconfig/compconfig/functions.py [hunk 3]
```diff
@@ -32,7 +37,8 @@ def get_inputs(meta_params_dict):
     params = Specifications()
     spec = params.specification(
         meta_data=True,
-        year=meta_params.start_year
+        serializable=True,
+        year=meta_params.year
     )
     return meta_params.specification(meta_data=True), {"ccc": spec}
```
Comment (hdoupe) [inline compconfig/compconfig/functions.py]:
```suggestion
    return meta_params.specification(meta_data=True, serializable=True), {"ccc": spec}
```

File: compconfig/compconfig/functions.py (hunk 2 above)
Comment (hdoupe) [reply compconfig/compconfig/functions.py]: Suggested one more change. Hopefully, that resolves the issue.

File: compconfig/compconfig/functions.py (hunk 3 above)
Comment (jdebacker) [reply compconfig/compconfig/functions.py]: This worked!  Thanks!

**Comment by jdebacker** [comment]
Tests appear to all pass.  Only issue I had is that I can't run locally with AWS PUF as datasource.  I guess I need to configure other things for that.  So I set the default data source to be the CPS.

**Comment by hdoupe** [comment]
Success(ish)!
Screenshot from 2019-05-08 18-32-56

**Comment by hdoupe** [comment]
I got this error when I tried to run it:

Screenshot from 2019-05-08 18-34-31

**Comment by hdoupe** [comment]
It looks like this is the problem:
```python
In [1]: from compconfig import functions                                                                                      

In [2]: mp = functions.MetaParams()                                                                                           

In [3]: mp.year                                                                                                               
Out[3]: array(2019)

In [4]: isinstance(mp.year, int)                                                                                              
Out[4]: False

In [5]: mp.year.tolist()                                                                                                      
Out[5]: 2019

In [6]: isinstance(mp.year.tolist(), int)                                                                                     
Out[6]: True

In [7]:                                                                                                                       
```

Tax-Calculator checks [if `start_year` is an `int`.](https://github.com/PSLmodels/Tax-Calculator/blob/master/taxcalc/records.py#L177-L182) ParamTools converts everything to an array from the list of dicts style, including non-list values like this. I consider this a bug and will fix this in ParamTools. In the mean time, you can call `tolist()` to convert the value to an `int`.

**Comment by jdebacker** [comment]
@hdoupe Ok, I was thinking it might be an error in that I was passing `start_year=None` to Tax-Calc from `get_taxcalc_rates.py`.  I updated to pass the default records year.  Let me know if that is not necessary.

**Comment by hdoupe** [comment]
I don't think it's necessary. A `start_year` value will be passed, but I think it's just not the right type.

**Comment by jdebacker** [comment]
@hdoupe Ok - but give it a shot with these latest changes and let me know if it works.

File: compconfig/compconfig/functions.py [hunk 4]
```diff
@@ -55,11 +69,15 @@ def run_model(meta_param_dict, adjustment):
     '''
     meta_params = MetaParams()
     meta_params.adjust(meta_param_dict)
-    params = Specifications(year=meta_params.start_year)
+    if meta_params.data_source == "PUF":
+        data = retrieve_puf(AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY)
+    else:
+        data = "cps"
+    params = Specifications(year=meta_params.year, call_tc=True, data=data)
```
Comment (hdoupe) [inline compconfig/compconfig/functions.py]: I think this needs to be converted to the `int` type via [`np.tolist`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.tolist.html):
```suggestion
    params = Specifications(year=meta_params.year.tolist(), call_tc=True, data=data)
```

**Comment by hdoupe** [comment]
I got the same error as before. I think the issue is that the year is some kind of numpy type instead of a Python `int` type.

File: compconfig/compconfig/functions.py [hunk 5]
```diff
@@ -55,11 +69,15 @@ def run_model(meta_param_dict, adjustment):
...
     params.adjust(adjustment["ccc"])
     assets = Assets()
     calc1 = Calculator(params, assets)
-    params2 = Specifications(year=meta_params.start_year)
+    params2 = Specifications(year=meta_params.year)
```
Comment (hdoupe) [inline compconfig/compconfig/functions.py]: Same as above:
```suggestion
    params2 = Specifications(year=meta_params.year.tolist())
```

**Comment by hdoupe** [comment]
Alright, it works:

Screenshot from 2019-05-08 22-18-10

Just had to install from source and convert a dataframe to a csv. Here's the diff: https://github.com/jdebacker/Cost-of-Capital-Calculator/compare/comp2...hdoupe:comp2

Also, the `tolist` changes weren't necessary. I think the main issue was that there were some updates in this branch that weren't being used because the conda version of CCC was still being installed. Thanks for bearing with me on this!

**Comment by hdoupe** [comment]
Just to make it clear, the screen shots are from a local instance of COMP.

**Comment by jdebacker** [comment]
@hdoupe That's awesome!  

Let me clean up this branch, merge changes, then work on getting the CCC package to build.

**Comment by hdoupe** [comment]
Sweet, we change the CCC install back to using conda when the new CCC package is released.

**Pull request merged**